activerecord/test/cases/associations/eager_test.rb

```diff
@@ -1487,6 +1487,21 @@ def test_eager_load_multiple_associations_with_references
     assert_equal posts(:welcome), post
   end
 
+  test "preloading of non-relations" do
+    class SpecialPostPreloader < ActiveRecord::Associations::Preloader::HasMany
+      def initialize(authors)
+        super(Post, authors, Author.reflect_on_association(:posts), ActiveRecord::Relation.new(nil, nil, nil))
+      end
+
+      def records_for(author_ids)
+        # to_a on an actual relation is called to simulate a real case
+        Post.where(author_id: author_ids).to_a
+      end
+    end
+
+    SpecialPostPreloader.new(Author.first(10)).run(nil)
```

Comment:
This is a no-go.

I could live with a test that relied on subclassing the internal preloader class and then did something to invoke it via public API... but this is a very long way from reasonable. If the use case is really "inherit from internal class, then construct and invoke it directly, just to reuse some of its behaviour", then I have no interest in such a test, nor in making any implication that this won't be broken in the future: it absolutely will.

Having now looked at this use case, I'm not so sure we should change anything. I'm honestly pretty offended that you've opened this at all: at this point five people have spent time on this, and the issue you're seeking to "fix" is that your monkey-patch of deep internals stopped working because the internals changed?!